### PR TITLE
Add token source enum

### PR DESCRIPTION
**Pull request by neha-bhargava**
#1728

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/AuthenticationResultMetadata.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+{
+    /// <summary>
+    /// Contains metadata of the authentication result.
+    /// </summary>
+    public class AuthenticationResultMetadata
+    {
+        /// <summary>
+        /// Default constructor for the class AuthenticationResultMetadata
+        /// </summary>
+        public AuthenticationResultMetadata()
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/AuthenticationResultMetadata.cs:17]: I don't see any reason for this object to be mutable (i.e. have a default ctor and a property setter). Immutable objects are  much easier to work with - see https://www.linkedin.com/pulse/20140528113353-16837833-6-benefits-of-programming-with-immutable-objects-in-java/ ... Mutability for this particular object is not very important, but it's good practice to be immutable by default.

File: src/client/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj [hunk 2]
```diff
@@ -27,5 +27,11 @@
   <ItemGroup>
     <ReferenceAssemblyProjectReference Include="..\Microsoft.Identity.Client.Ref\Microsoft.Identity.Client.Ref.csproj" />
   </ItemGroup>
+  <ItemGroup>
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj:30]: Please undo the changes to this file. Visual Studio gets a little confused sometimes when adding files, but the general pattern is that we include all .cs files by default, so this is not needed.

File: src/client/Microsoft.Identity.Client/TokenSource.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Client
+{
+    /// <summary>
+    /// Specifies the source of the access and Id tokens in the authentication result.
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/TokenSource.cs:10]: I made some changes to the comments - please have a look. I though GitHub would "propose changes"... I guess I'm not a GitHub power user...

File: tests/Microsoft.Identity.Test.Integration.net45/HeadlessTests/ConfidentialClientIntegrationTests.cs [hunk 4]
```diff
@@ -133,6 +133,7 @@ public async Task ConfidentialClientWithCertificateTestAsync()
             appCacheRecorder.AssertAccessCounts(1, 1);
             Assert.IsTrue(appCacheRecorder.LastAfterAccessNotificationArgs.IsApplicationCache);
             Assert.IsTrue(appCacheRecorder.LastAfterAccessNotificationArgs.HasTokens);
+            Assert.AreEqual(authResult.AuthenticationResultMetadata.TokenSource, TokenSource.IdentityProvider);
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Integration.net45/HeadlessTests/ConfidentialClientIntegrationTests.cs:136]: nit: in the Assert methods, the "expected" value is the first value.

**Comment by jennyf19** [comment]
> using System;

license info.

---
Refers to: src/client/Microsoft.Identity.Client/TokenSource.cs:1 in f6dd77f. [](commit_id = f6dd77f5f45237e0f31cab33808b29056b0ab67a, deletion_comment = False)

File: src/client/Microsoft.Identity.Client/TokenSource.cs (hunk 3 above)
Comment (jennyf19) [reply src/client/Microsoft.Identity.Client/TokenSource.cs:10]: lol. 

---
In reply to: [453897525](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1933#discussion_r453897525) [](ancestors = 453897525)

File: src/client/Microsoft.Identity.Client/TokenSource.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        None = 0,
+        /// <summary>
+        /// The source of access and Id token is MSAL's cache.
+        /// </summary>
+        Cache,
+        /// <summary>
+        /// The source of the access and Id token is Identity Provider - Azure Active Directory (AAD), ADFS or AAD B2C.
+        /// </summary>
+        IdentityProvider,
+        /// <summary>
```
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/TokenSource.cs:26]: nit: would you want to separate b2c from AAD and ADFS or does it matter? Should it be called `AuthorizationServer`? When i think of an Identity provider, I think of Google or FB or Microsoft...really just a question

File: tests/Microsoft.Identity.Test.Integration.net45/HeadlessTests/SilentAuthTests.cs [hunk 6]
```diff
@@ -155,6 +155,7 @@ public async Task SilentAuth_TokenCacheRemainsPersistent_Async()
             Assert.IsFalse(at1.Equals(at2, System.StringComparison.InvariantCultureIgnoreCase));
             Assert.IsFalse(at1.Equals(at3, System.StringComparison.InvariantCultureIgnoreCase));
             Assert.IsFalse(at2.Equals(at3, System.StringComparison.InvariantCultureIgnoreCase));
+            Assert.AreEqual(TokenSource.IdentityProvider, authResult.AuthenticationResultMetadata.TokenSource);
         }
```
Comment (jennyf19) [inline tests/Microsoft.Identity.Test.Integration.net45/HeadlessTests/SilentAuthTests.cs:159]: can you assert earlier, like after line 133, that the TokenSource is `TokenSource.Cache`...I guess finding one that does AT silent and pulls from the cache. I would just add this check in all the tests, what about the interactive tests and broker tests?

File: src/client/Microsoft.Identity.Client/TokenSource.cs (hunk 3 above)
Comment (jmprieur) [reply src/client/Microsoft.Identity.Client/TokenSource.cs:10]: You need to use the +/- icon to proposed suggestions, @bgavrilMS 
image

File: src/client/Microsoft.Identity.Client/TokenSource.cs (hunk 5 above)
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/TokenSource.cs:26]: The goal of this small improvement is for developers, especially on Identity.Web, to monitor their token cache hit ratio. There are other proposals around providing more details such as "which broker (name and version) was used?", but will treat them separately. 

Not sure about the naming, I proposed the OAuth2 term. @jmprieur - what do you think?

File: src/client/Microsoft.Identity.Client/TokenSource.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+{
+    /// <summary>
+    /// Specifies the source of the access and Id tokens in the authentication result.
+    /// </summary>
+    public enum TokenSource
+    {
+        /// <summary>
+        /// The source of access and Id token is not set.
+        /// </summary>
+        None = 0,
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/TokenSource.cs:17]: It looks like you are using this value for internal logic, but it does not have any meaning in the public API. There is always a source for the token. Let's not expose this.

File: src/client/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs [hunk 8]
```diff
@@ -210,7 +210,8 @@ protected async Task<AuthenticationResult> CacheTokenResponseAndCreateAuthentica
                 atItem, 
                 idtItem, 
                 AuthenticationRequestParameters.AuthenticationScheme,
-                AuthenticationRequestParameters.RequestContext.CorrelationId);
+                AuthenticationRequestParameters.RequestContext.CorrelationId,
+                msalTokenResponse.TokenSource == TokenSource.None? TokenSource.IdentityProvider : msalTokenResponse.TokenSource);
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs:214]: Why can't msalTokenResponse.TokenSource always have the correct value?

File: src/client/Microsoft.Identity.Client/TokenSource.cs [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        /// </summary>
+        Cache,
+        /// <summary>
+        /// The source of the access and Id token is Identity Provider - Azure Active Directory (AAD), ADFS or AAD B2C.
+        /// </summary>
+        IdentityProvider,
+        /// <summary>
+        /// The source of the access and Id token is a broker application - Authenticator or Company Portal. Brokers are supported only on Android and iOS.
+        /// </summary>
+        Broker
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/TokenSource.cs:29]: There are no unit tests that cover this. While the brokers are really only implemented in mobile code, which is not testable, it would be valuable to have a test that proves that the TokenSource value is indeed copied from MsalTokenSource to AuthenticationResult. I believe there are (a rather few) tests that mock IBroker, such as https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/ad7970c3904d25662e5d756d076f777b0ed1867a/tests/Microsoft.Identity.Test.Unit.net45/RequestsTests/BrokerRequestTests.cs#L183 - you can use this approach to unit test this.

File: src/client/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs (hunk 8 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs:214]: There are methods in the MsalTokenResponse class to build response for broker. So I am setting the value of token source to broker there. https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1933/files#diff-ea5df3781a2dc2119f53390344fbfbd7R136

This method will either have the token source as Broker or Identity Provider. So instead I can check if Broker is set and otherwise set Identity Provider.

File: src/client/Microsoft.Identity.Client/TokenSource.cs (hunk 7 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/TokenSource.cs:17]: I was using it for internal logic but also I wanted to keep default to some other value in case the token source is not set or missed in some flow. Maybe rename None to Unidentified to make it more relevant?

File: src/client/Microsoft.Identity.Client/TokenSource.cs (hunk 9 above)
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/TokenSource.cs:29]: Actually this test is a better fit: https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/029e6031a1bccf88c5e31154712fcaa60a1fb4dd/tests/Microsoft.Identity.Test.Unit.net45/RequestsTests/InteractiveRequestOrchestrationTests.cs#L106

**Comment by trwalke** [comment]
>             return await _silentRequest.CacheTokenResponseAndCreateAuthenticationResultAsync(msalTokenResponse).ConfigureAwait(false);

If the access token expires the refresh token is used to refresh the token and acquire a new access token. Thus, the token is actually coming from AAD in this case. are you accounting for this scenario somewhere?

---
Refers to: src/client/Microsoft.Identity.Client/Internal/Requests/Silent/SilentClientAuthStrategy.cs:119 in 0e4447b. [](commit_id = 0e4447bc28955687d648b04dccfe7d05293b59bb, deletion_comment = False)

**Comment by trwalke** [comment]
>             return await _silentRequest.CacheTokenResponseAndCreateAuthenticationResultAsync(msalTokenResponse).ConfigureAwait(false);

Also, is force refresh being handled? it should also be from AAD during silent calls

---
In reply to: [659063735](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1933#issuecomment-659063735) [](ancestors = 659063735)

---
Refers to: src/client/Microsoft.Identity.Client/Internal/Requests/Silent/SilentClientAuthStrategy.cs:119 in 0e4447b. [](commit_id = 0e4447bc28955687d648b04dccfe7d05293b59bb, deletion_comment = False)

**Comment by trwalke** [comment]
> > ```
> >         return await _silentRequest.CacheTokenResponseAndCreateAuthenticationResultAsync(msalTokenResponse).ConfigureAwait(false);
> > ```
> 
> If the access token expires the refresh token is used to refresh the token and acquire a new access token. Thus, the token is actually coming from AAD in this case. are you accounting for this scenario somewhere?
> 
> Refers to: src/client/Microsoft.Identity.Client/Internal/Requests/Silent/SilentClientAuthStrategy.cs:119 in [0e4447b](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/commit/0e4447bc28955687d648b04dccfe7d05293b59bb). [](commit_id = [0e4447b](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/commit/0e4447bc28955687d648b04dccfe7d05293b59bb), deletion_comment = False)

File: tests/Microsoft.Identity.Test.Integration.net45/HeadlessTests/SilentAuthTests.cs (hunk 6 above)
Comment (bgavrilMS) [reply tests/Microsoft.Identity.Test.Integration.net45/HeadlessTests/SilentAuthTests.cs:159]: Or after line 145, that AcquireTokenSilent gets a token from teh cache.

Comment (neha-bhargava) [reply tests/Microsoft.Identity.Test.Integration.net45/HeadlessTests/SilentAuthTests.cs:159]: That one is withForceRefresh(True) so it would not get the token from the cache. Right?

File: src/client/Microsoft.Identity.Client/TokenSource.cs (hunk 7 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/TokenSource.cs]: Removed.

File: tests/Microsoft.Identity.Test.Integration.net45/HeadlessTests/SilentAuthTests.cs (hunk 6 above)
Comment (neha-bhargava) [reply tests/Microsoft.Identity.Test.Integration.net45/HeadlessTests/SilentAuthTests.cs:159]: Added in the other method where force refresh was false.

Comment (bgavrilMS) [reply tests/Microsoft.Identity.Test.Integration.net45/HeadlessTests/SilentAuthTests.cs:159]: the assertion at line 158 is correct - force_refresh means "do not look for an AT in the cache".
